network-libp2p/src/discovery/behaviour.rs

```diff
@@ -318,11 +304,9 @@ impl NetworkBehaviour for Behaviour {
                         }));
                 }
             }
-            HandlerOutEvent::ObservedAddresses { observed_addresses } => {
-                for address in observed_addresses {
-                    self.events
-                        .push_back(ToSwarm::NewExternalAddrCandidate(address));
-                }
+            HandlerOutEvent::ObservedAddress { observed_address } => {
+                self.events
+                    .push_back(ToSwarm::NewExternalAddrCandidate(observed_address));
             }
             HandlerOutEvent::Update => self.events.push_back(ToSwarm::GenerateEvent(Event::Update)),
             HandlerOutEvent::Error(_) => self.events.push_back(ToSwarm::CloseConnection {
```

Comment:
If I understand correctly this is where the `StateTransitionTimeout` would end up. I may have missed it, but I think some logging would be nice here.